lxd/device/device.go

```diff
@@ -0,0 +1,74 @@
+package device
+
+import (
+	"fmt"
+)
+
+// ErrUnsupportedDevType is the error that occurs when an unsupported device type is created.
+var ErrUnsupportedDevType = fmt.Errorf("Unsupported device type")
+
+// Config represents user defined persistent config for a device.
+type Config map[string]string
+
+// VolatileConfig represents LXD defined persistent config for a device.
+type VolatileConfig map[string]string
+
+// RunConfig represents LXD defined run-time config used for device setup.
+type RunConfig map[string]string
+
+// Device represents a device that can be added to an instance.
+type Device interface {
+	// Start peforms any host-side configuration required to setup the device for the instance.
+	// This can be when a device is plugged into a running instance or the instance is starting.
+	// VolatileConfig from previous instantiations of this device is required.
+	// It returns run-time configuration necessary for configuring the instance with the new
+	// device and any changes to VolatileConfig that need to be persisted.
+	Start(VolatileConfig) (RunConfig, VolatileConfig, error)
+
+	// Stop performs any host-side cleanup required when a device is removed from an instance,
+	// either due to unplugging it from a running instance or instance is being shutdown.
+	// VolatileConfig from previous instantiations of this device is required.
+	// It return sany changes to VolatileConfig that need to be persisted.
+	Stop(VolatileConfig) (VolatileConfig, error)
+
+	// setConfig stores the contents of instanceType and Config into device.
+	setConfig(string, Config)
```

Comment:
Having an un-exported function in an interface looks odd, if it's internal then it shouldn't be there and if it's external, it should be exported.

Reply:
@stgraber this and the validate() function below makes this a so-called "sealed interface", meaning it cannot be implemented outside of the `devices` package. I'm not sure whether this is desirable for us or not, but my intention was to limit the scope of what is exported (including its implementations) early only. However I agree it is not immediately clear why this is the case. 

Let me have a think about ways around this that don't also allow these devices to be instantiated without valid config being checked.

Reply:
See https://github.com/lxc/lxd/pull/5900#discussion_r299902271